asf_levies_model/consumers.py

```diff
@@ -0,0 +1,460 @@
+    def consumer_from_dataframe(
+        cls,
+        df: pd.DataFrame,
+        row: int,
+        name_col: str,
+        archetype_col: str,
+        net_annual_income_col: str,
+        main_heating_fuel_col: str,
+        gas_consumption_col: str,
+        electricity_consumption_col: str,
+        gas_tariff: object,
+        electricity_tariff: object,
+        net_income_decile_col: str = None,
+        unmetered_fuel_spend_col: str = None,
+        scheme_eligible_col: str = None,
+        size_col: str = None,
+        unit_converter: float = 1,
+    ) -> "Consumer":
+        """Creates a Consumer instance from dataframe input.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            Dataframe where each row corresponds to a consumer profile and columns describe (at least) name, net annual, income, main heating fuel, annual gas consumption, annual electricity consumption.
+        row : int
+            Row number in dataframe for consumer profile to instantiate.
+        name_col : str
+            Name of the dataframe column with unique consumer name (str).
+        archetype_col : str
+            Name of the dataframe column with consumer profile category (str).
+        net_annual_income_col : str
+            Name of the dataframe column with net annual income amount (float).
+        main_heating_fuel_col : str
+            Name of the dataframe column with main heating fuel description (str).
+        gas_consumption_col : str
+            Name of the dataframe column with annual gas consumption amount (float).
+        electricity_consumption_col : str
+            Name of the dataframe column with annual electricity consumption amount (float).
+        gas_tariff : object
+            Gas tariff object for consumer bill to be calculated with.
+        electricity_tariff : object
+            Electricity tariff object for consumer bill to be calculated with.
+        net_income_decile_col : str, optional
+            Name of the dataframe column with income decile (int), by default None.
+        unmetered_fuel_spend_col : str, optional
+            Name of the dataframe column with annual unmetered fuel spend amount (float), by default None.
+        scheme_eligible_col : str, optional
+            Name of the dataframe column with scheme eligibility (bool), by default None.
+        size_col : str, optional
+            Name of the dataframe column with the number of households in consumer profile if being considered, by default 0.
+        unit_converter : float, optional
+            Value to divide electricity and gas consumption values from dataframe by to convert to MWh, by default 1.
+
+        Returns
+        -------
+        Consumer
+            Consumer instance.
+        """
+
+        def safe_get(col_name, default=None):
+            """Helper function to retrieve column value if column name is provided."""
+            return df.loc[row, col_name] if col_name else default
+
```

Comment:
There's no reason to do this, instead I recommend `df.iloc[row]` to get a pandas series of the desired row, then `Series.get()` to get the column value of interest, with an optional default.